src/util/time.cpp

```diff
@@ -9,39 +9,151 @@
 #endif
 
 #include "util/time.h"
+#include "sync.h"
 
 #include <chrono>
 #include <boost/date_time/posix_time/posix_time.hpp>
 #include <boost/thread.hpp>
 
-using namespace std;
+RecursiveMutex cs_clock;
+static CClock* zcashdClock = SystemClock::Instance();
 
-static int64_t nMockTime = 0; //!< For unit testing
+const CClock* GetNodeClock() {
+    return zcashdClock;
+}
 
-int64_t GetTime()
-{
-    if (nMockTime) return nMockTime;
+int64_t GetTime() {
+    return zcashdClock->GetTime();
+}
 
-    return time(NULL);
+int64_t GetTimeMillis() {
+    return zcashdClock->GetTimeMillis();
 }
 
-void SetMockTime(int64_t nMockTimeIn)
-{
-    nMockTime = nMockTimeIn;
+int64_t GetTimeMicros() {
+    return zcashdClock->GetTimeMicros();
 }
 
-int64_t GetTimeMillis()
-{
+void SystemClock::SetGlobal() {
+    zcashdClock = SystemClock::Instance();
+}
+
+int64_t SystemClock::GetTime() const {
+    return std::chrono::duration_cast<std::chrono::seconds>(
+            std::chrono::system_clock::now().time_since_epoch()).count();
+}
+
+int64_t SystemClock::GetTimeMillis() const {
     return std::chrono::duration_cast<std::chrono::milliseconds>(
             std::chrono::system_clock::now().time_since_epoch()).count();
 }
 
-int64_t GetTimeMicros()
-{
+int64_t SystemClock::GetTimeMicros() const {
     return std::chrono::duration_cast<std::chrono::microseconds>(
             std::chrono::system_clock::now().time_since_epoch()).count();
 }
 
+void FixedClock::SetGlobal() {
+    LOCK(cs_clock);
+    zcashdClock = FixedClock::Instance();
```

Comment:
I'm now confused about what the intended locking was for `SetGlobal` (which per our discussion in PR gardening I thought was to not lock):
- `SystemClock` doesn't lock anywhere.
- `FixedClock` locks everwhere.
- `OffsetClock` locks everwhere _except_ `OffsetClock::SetGlobal`.

Reply:
Oh. Hm. FixedClock::SetGlobal wasn't supposed to lock. I'm not sure how I missed that. Will fix.